bee-protocol/src/worker/transaction/hasher.rs

```diff
@@ -0,0 +1,188 @@
+            cache,
+            ..
+        } = self.project();
+
+        // We loop until we have `BATCH_SIZE` transactions or `stream.poll_next(cx)` returns
+        // pending.
+        loop {
+            let batch_size = hasher.len();
+            // If we already have `BATCH_SIZE` transactions, we are ready.
+            if batch_size == BATCH_SIZE {
+                return Poll::Ready(Some(BATCH_SIZE));
+            }
+            // Otherwise we need to check if there is transactions inside `stream` that we could
+            // include in the current batch.
+            match receiver.as_mut().poll_next(cx) {
+                Poll::Pending => {
+                    return if batch_size == 0 {
+                        // If the stream is not ready yet and the current batch is empty we have to
+                        // wait. Otherwise, we would end up hashing an empty batch.
+                        Poll::Pending
+                    } else {
+                        // If the stream is not ready yet, but we have transactions in the batch,
+                        // we can process them instead of waiting.
+                        Poll::Ready(Some(batch_size))
+                    };
+                }
+                Poll::Ready(Some(event)) => {
+                    // If the transaction was already received, we skip it and poll again.
+                    if !cache.insert(&event.transaction.bytes) {
+                        debug!("Transaction already received.");
+                        Protocol::get().metrics.known_transactions_inc();
+                        continue;
+                    }
+                    // Given that the current batch has less than `BATCH_SIZE` transactions. We can
+                    // add the transaction in the current event to the batch.
+                    let transaction_bytes = uncompress_transaction_bytes(&event.transaction.bytes);
+
+                    let trits = Trits::<T5B1>::try_from_raw(cast_slice(&transaction_bytes), 8019)
+                        .unwrap()
+                        .to_buf::<T5B1Buf>()
+                        .encode::<T1B1Buf>();
+
+                    hasher.add(trits);
+                    events.push(event);
+                    // If after adding the transaction to the batch, its size is `BATCH_SIZE` we
+                    // are ready to hash.
+                    if batch_size == BATCH_SIZE - 1 {
+                        return Poll::Ready(Some(BATCH_SIZE));
+                    }
+                }
+                Poll::Ready(None) => {
+                    return if batch_size == 0 {
+                        // If the stream ended and the current batch is empty, we have nothing else
+                        // to hash.
+                        Poll::Ready(None)
+                    } else {
+                        // IF the stream ended but we have some transactions in the batch, we must
+                        // hash them before being done.
```

Comment:
Should we really ?

Reply:
Well we already have them, the hasher can hit 88 mph so I don't think this will slowdown shutdown in any relevant way.

Reply:
It's just that it means we already started shutting down other workers as well sot it's going to trigger things that will fail

Reply:
makes total sense.